File: 100 Same Tree/100.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque
class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q:
            return True
        elif not p or not q:
            return False
        elif p.val != q.val:
            return False
        else:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

    def isSameTree2(self, p: TreeNode, q: TreeNode) -> bool:
        stack = deque()
        stack.append(p)
        stack.append(q)
        # can't write like this
        # while stack[0] -> this first element is not None, it not equal to the !isEmpty()
        while stack:
            right = stack.pop()
            left = stack.pop()
            if not right or not left:
                if right != left:
                    return False
            elif left.val != right.val:
                return False
            else:
                stack.append(left.left)
                stack.append(right.left)
                stack.append(left.right)
                stack.append(right.right)
        return True

File: 100 Same Tree/100.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function(p, q) {
    if (!p && !q) {
        return true;
    } else if (!q || !p) {
        return false;
    } else if (p.val != q.val) {
        return false;
    } else {
        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right)
    }
};

var isSameTree2 = function(p, q) {
    let stack = []
    stack.push(p)
    stack.push(q)
    while (stack.length) {
        console.log(stack.length)
        let right = stack.pop()
        let left = stack.pop()
        if (!left || !right) {
            if (left != right) {
                return false;
            }
        } else if (left.val != right.val) {
            return false;
        } else {
            stack.push(left.left)
            stack.push(right.left)
            stack.push(left.right)
            stack.push(right.right)
        }
    }
    return true;
};